test_scripts/API/VehicleData/common.lua

```diff
@@ -636,4 +543,675 @@ function m.setAppVersion(pParamVersion, pOperator)
+      tc.graph[tc.paramId].data = { invalidValue }
+      local params = getParamsFuncMap.INVALID[rpcType](tc.graph)
+      table.insert(tests, {
+          name = "Param_" .. ah.getFullParamName(tc.graph, tc.paramId) .. "_" .. invalidValue,
+          params = params
+        })
+    end
+  end
+  return tests
+end
+
+--[[ @getEnumItemsTests: Generate tests for ENUM_ITEMS test type
+--! @parameters: none
+--! @return: table with tests
+--]]
+local function getEnumItemsTests()
+  local tests = {}
+  local dataTypes = { ah.dataType.ENUM.type }
+  local tcs = createTestCases(ah.apiType.HMI, rpcType, m.rpcHMIMap[rpc],
+    m.isMandatory.ALL, m.isArray.ALL, m.isVersion.ALL, dataTypes)
+  for _, tc in pairs(tcs) do
+    for _, item in pairs(tc.graph[tc.paramId].data) do
+      local tcUpd = utils.cloneTable(tc)
+      tcUpd.graph[tc.paramId].data = { item }
+      table.insert(tests, {
+          name = "Param_" .. ah.getFullParamName(tc.graph, tc.paramId) .. "_" .. item,
+          params = getParamsFuncMap.VALID[rpcType](tcUpd.graph)
+        })
+    end
+  end
+  return tests
+end
+
+--[[ @getBoolItemsTests: Generate tests for BOOL_ITEMS test type
+--! @parameters: none
+--! @return: table with tests
+--]]
+local function getBoolItemsTests()
+  local tests = {}
+  local dataTypes = { ah.dataType.BOOLEAN.type }
+  local tcs = createTestCases(ah.apiType.HMI, rpcType, m.rpcHMIMap[rpc],
+    m.isMandatory.ALL, m.isArray.ALL, m.isVersion.ALL, dataTypes)
+  for _, tc in pairs(tcs) do
+    for _, item in pairs({ true, false }) do
+      local tcUpd = utils.cloneTable(tc)
+      tcUpd.graph[tc.paramId].data = { item }
+      table.insert(tests, {
+          name = "Param_" .. ah.getFullParamName(tc.graph, tc.paramId) .. "_" .. tostring(item),
+          params = getParamsFuncMap.VALID[rpcType](tcUpd.graph)
+        })
+    end
+  end
+  return tests
+end
+
+--[[ @getVersionTests: Generate tests for PARAM_VERSION test type
+--! @parameters: none
+--! @return: table with tests
+--]]
+local function getVersionTests()
+  local tests = {}
+  local tcs = createTestCases(ah.apiType.MOBILE, ah.eventType.REQUEST, rpc,
+    m.isMandatory.ALL, m.isArray.ALL, m.isVersion.YES, {})
+  for _, tc in pairs(tcs) do
+    local name = tc.graph[tc.paramId].name
+    if vdParams[name] then
+      table.insert(tests, {
+          param = tc.graph[tc.paramId].name,
+          version = tc.graph[tc.paramId].since
+        })
+    end
+  end
+  return tests
+end
+
+--[[ @getValidRandomAllTests: Generate tests for VALID_RANDOM_ALL test type
+--! @parameters: none
+--! @return: table with tests
+--]]
+local function getValidRandomAllTests()
+  local tests = {}
+  local graph = ah.getGraph(ah.apiType.HMI, rpcType, m.rpcHMIMap[rpc])
+  local function getParamId(pGraph, pName)
+    for k, v in pairs(pGraph) do
+      if v.parentId == nil and v.name == pName then return k end
+    end
+    return nil
+  end
+  local paramId = getParamId(graph, paramName)
```

Comment:
Function `getParamId()` allows to find `id` in the graph of the root VD parameter by it's `name`.
Then this `id` is used to get the branch of the graph which corresponds to only single VD parameter.

Reply:
It still seems to me that these two would be identical, based on the logic (both of these would be the `key` for the parameter).

Can you give me an example where `paramId` is different than `paramName`?

Reply:
Seems I got the question.

Initialy in a previous approach name of the parameter was used to identify a parameter.
It was Ok since only `root` level was verified.
However in a new approach it was found that can exist 2 parameters with the same `name` in a different branches of hierarchy.
E.g. 

1) `pressure`:
  - tirePressure
    - leftFront 
      - pressure 
    - rightFront
      - pressure

2) `col`
 - bodyInformation
   - doorStatuses
     - location
       - col
    - gateStatuses
      - location
        - col

In order to have unique identifier for each parameter (and sub-parameter) it was decided to move from `name` to `id`

Reply:
Alright, I can see the case where a different identifier would be necessary, but I still do not see how this logic accomplishes this:
```
local function getParamId(pGraph, pName)
    for k, v in pairs(pGraph) do
      if v.parentId == nil and v.name == pName then return k end
    end
    return nil
  end
```
all this seems to do is return the name of the param (if it exists at the root level). Is `k` in this case the full parameter path (ex. `tirePressure.leftFront.pressure`)? 

It seems to me that it is only the name (ex. `pressure`, which is already contained in `paramName`)

Reply:
I guess I need to explain what is the `graph` object.
Graph object in these tests is a table which represents hierarchy of all VD parameters (incl. sub-parameters).

`key` in this table is unique identifier of the VD parameter, this is an integer type value.
`value` in this table is a set of elements. Most important ones are: 
 - `parentId` - identifier of a parent (or nil for the root elements), this is also integer value.
 - `name` - name of the parameter
 - `<restrictions>`, such as `type`, `array`, `mandatory`, `since` etc. copied from API

`graph` object can be easily traversed on any level of hierarchy and is used to build tree of parameters and it's values.

Having that `getParamId()` function returns `Id` of the parameter (`Integer`) by it's name (`String`)

Reply:
Alright, I believe I understand now. However, this does somewhat tie into my point that this "graph" approach isn't particularly intuitive or clearly described. Using this approach just seems to complicate the logic as far as I can tell.